packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts

```diff
@@ -109,248 +111,149 @@ export class FilteringHierarchyDefinition implements HierarchyDefinition {
+                    return false;
+                  }
+                  if (!HierarchyNodeIdentifier.isInstanceNodeIdentifier(identifier)) {
+                    return false;
+                  }
+                  if (identifier.imodelKey && identifier.imodelKey !== this._imodelAccess.imodelKey) {
+                    return false;
+                  }
+                  if (identifier.className === rowInstanceKey.className) {
+                    return true;
+                  }
+                  return Promise.all([
+                    this._imodelAccess.classDerivesFrom(identifier.className, rowInstanceKey.className),
+                    this._imodelAccess.classDerivesFrom(rowInstanceKey.className, identifier.className),
+                  ]).then(([isDerivedFrom, isDerivedTo]) => isDerivedFrom || isDerivedTo);
+                },
+              });
+            })()
+          : undefined;
+      if (nodeExtraProps?.autoExpand) {
+        parsedNode.autoExpand = true;
       }
-
-      // We need to check if identifiers have the same id and imodelKey and are
-      // of the same class / derived class / is derived from class
-      const identifier = path[0];
-      if (identifier.id !== id) {
-        continue;
+      if (nodeExtraProps?.filtering) {
+        parsedNode.filtering = nodeExtraProps.filtering;
       }
-      if (HierarchyNodeIdentifier.isInstanceNodeIdentifier(identifier)) {
-        if (identifier.imodelKey && identifier.imodelKey !== this._imodelAccess.imodelKey) {
-          continue;
-        }
-        if (
-          identifier.className !== className &&
-          !(await this._imodelAccess.classDerivesFrom(identifier.className, className)) &&
-          !(await this._imodelAccess.classDerivesFrom(className, identifier.className))
-        ) {
-          continue;
-        }
-      }
-      if (!filteredChildrenIdentifierPaths) {
-        filteredChildrenIdentifierPaths = [];
-      }
-      if (path.length > 1) {
-        filteredChildrenIdentifierPaths.push(options ? { path: path.slice(1), options } : path.slice(1));
-      } else if (isFilterTarget) {
-        filterTargetOptions = HierarchyFilteringPath.mergeOptions(filterTargetOptions, options);
-      } else {
-        isFilterTarget = true;
-        filterTargetOptions = options;
-      }
-    }
-    return { filteredChildrenIdentifierPaths, filterTargetOptions, isFilterTarget };
+
+      return parsedNode;
+    };
   }
 
   public async defineHierarchyLevel(props: DefineHierarchyLevelProps): Promise<HierarchyLevelDefinition> {
     const sourceDefinitions = await this._source.defineHierarchyLevel(props);
-    const filteringProps = extractFilteringProps(this._nodeIdentifierPaths, props.parentNode);
-    if (!filteringProps) {
+
+    const filteringHelper = createHierarchyFilteringHelper(this._nodeIdentifierPaths, props.parentNode);
+    const childNodeFilteringIdentifiers = filteringHelper.getChildNodeFilteringIdentifiers();
```

Comment:
We are getting `childNodeFilteringIdentifiers` and later we are iterating over this array twice: once for `genericNodeDefinitions`, and once for `instanceNodeDefinitions`. Maybe we could split this array into two after checking if it's not undefined. In that case we would iterate only once over this array.

Reply:
I think you suggestion would slightly help if the split would divide the array in more-or-less equal parts. However, I predict that whenever we get many filtered node paths, absolute majority of them are going to be of one type. And in that case, iterating over it once per definition is better than iterating once to split + once per definition.